docs: rewrite description of zkevm-prover (#70)
kroma-network/kroma: specs/zkevm-prover.md

[g-state]: glossary.md#state
[g-zk-fault-proof]: glossary.md#zk-fault-proof
[g-mpt]: glossary.md#merkle-patricia-trie

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
- [ZKEVM Proofs](#zkevm-proofs)
- [The ZK Verifier Contract](#the-zk-verifier-contract)
- [Prover as an RPC Server](#prover-as-an-rpc-server)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

A Prover is responsible for generating proofs using the [Halo2 proving scheme](https://zcash.github.io/halo2/)

> for a target blocks. this proof provides claims regarding the legitimacy of block data and state transitions.

Reviewer: ```suggestion
for a target blocks. This proof provides claims regarding the legitimacy of block data and state transitions.
```

Author: fixed

> A prover plays a vital role in the challenge process, serving as a ke component of the challenger.

Reviewer: ```suggestion
A prover plays a vital role in the challenge process, serving as a key component of the challenger.
```

Author: fixed

A [ZK fault proof][g-zk-fault-proof] states that a [state][g-state] transition from `S` to `S'` is valid.
It sounds like there are no big differences from validity proofs. That's true. But the point is this is used
to prove the state transition `S` to `S''` is wrong by showing a valid state transition `S` to `S'`.

## ZKEVM Proofs

In the context of the EVM, which is regarded as a Turing machine, the generation of proofs requires demonstrating the
execution of opcode operations in arbitrary order. This proof serves as an evidence of computation for the EVM.
Additionally, validating the block data's integrity involves proving sub-operations such as
[Merkle Patricia Trie][g-mpt] and ECDSA verification.
To support the proof generation process for a target block, [zkevm-circuits] offer circuits capable of proving the
required sub-operations. These circuits provide the means to prove the integrity and validity of the block data.
By combining these circuits, a prover completes the overall validity proof for the target block.

See [zkevm-specs] for details about the statements claimed by the zk-proof.

## The ZK Verifier Contract

> The proofs generated by the prover can be verified through a verifier contract that includes the following interface.

Reviewer: ```suggestion
The proof generated by a prover can be verified through the verifier contract that includes the following interface.
```

Author: fixed

[g-state]: glossary.md#state
[g-zk-fault-proof]: glossary.md#zk-fault-proof
[g-mpt]: glossary.md#merkle-patricia-trie

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
- [ZKEVM Proofs](#zkevm-proofs)
- [The ZK Verifier Contract](#the-zk-verifier-contract)
- [Prover as an RPC Server](#prover-as-an-rpc-server)
- [Summary of Definitions](#summary-of-definitions)
  - [Constants](#constants)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

> The Prover is responsible for generating proofs using the [Halo2 proving scheme](https://zcash.github.io/halo2/)

Author: applied

> for target blocks. These proofs provide claims regarding the legitimacy of block data and state transitions.

Reviewer: ```suggestion
for a target block. This proof provides claims regarding the legitimacy of block data and state transitions.
```

Author: applied

> And the prover is main feature of the challenge process as a component of challenger.

Author: applied

> [ZK fault proof][g-zk-fault-proof] states that a [state][g-state] transition from `S` to `S'` is valid.

Author: applied

It sounds like there are no big differences from validity proofs. That's true. But the point is this is used
to prove the state transition `S` to `S''` is wrong by showing a valid state transition `S` to `S'`.

## ZKEVM Proofs

In the context of the EVM, which is regarded as a Turing machine, the generation of proofs requires demonstrating the

> execution of Opcode operations in arbitrary order. These proofs serve as evidence of computation for the EVM.

Author: applied

Additionally, validating the block data's integrity involves proving sub-operations such as
[Merkle Patricia Trie][g-mpt] and ECDSA verification.

> To support the proof generation process for target blocks, [zkevm-circuits] offer circuits capable of proving the

Reviewer: ```suggestion
To support the proof generation process for a target block, [zkevm-circuits] offer circuits capable of proving the
```

Author: applied

required sub-operations. These circuits provide the means to prove the integrity and validity of the block data.

> By combining these circuits, the Prover completes the overall validity proof for the target block.

Reviewer: ```suggestion
By combining these circuits, a prover completes the overall validity proof for a target block.
```

Author: applied

See [zkevm-specs] for details about the statements claimed by the zk-proof.

## The ZK Verifier Contract

> The proofs generated by the Prover can be verified through a Verifier contract that includes the following interface.

Reviewer: ```suggestion
The proofs generated by a prover can be verified through a verifier contract that includes the following interface.
```

Author: applied

> The verification in challenge process is implemented using the `verify` function provided by the Verifier contract.

Reviewer: ```suggestion
The verification in a challenge process is implemented using the `verify` function provided by the verifier contract.
```

Author: applied

> <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

Reviewer 2: Could you apply markdownlint?

Author: applied

**Table of Contents**

> - [zkEVM Prover](#zkevm-prover)
>   - [Overview](#overview)
>   - [ZKEVM Proofs](#zkevm-proofs)
>   - [The ZK Verifier Contract](#the-zk-verifier-contract)
>   - [Prover as an RPC Server](#prover-as-an-rpc-server)
>     - [Protobuf](#protobuf)
>   - [Summary of Definitions](#summary-of-definitions)
>     - [Constants](#constants)

Reviewer 2: ```suggestion
- [Overview](#overview)
- [ZKEVM Proofs](#zkevm-proofs)
- [The ZK Verifier Contract](#the-zk-verifier-contract)
- [Prover as an RPC Server](#prover-as-an-rpc-server)
  - [Protobuf](#protobuf)
- [Summary of Definitions](#summary-of-definitions)
  - [Constants](#constants)
```

Author: applied.

Author: Since this toc was generated by doctoc, so it can not be modified. Do you have any idea to change it?

Reviewer 2: Oh I see it. I just suggested it cause the format of toc is different with other docs.

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

The Prover is responsible for generating proofs using the [Halo2 proving scheme](https://zcash.github.io/halo2/)
for target blocks. These proofs provide claims regarding the legitimacy of block data and state transitions.
And the prover is main feature of the challenge process as a component of challenger.

> [ZK fault proof][g-zk-fault-proof] that states that a [state][g-state] transition from `S` to

Reviewer 2: ```suggestion
[ZK fault proof][g-zk-fault-proof] states that a [state][g-state] transition from `S` to
```

Author: applied

`S'` is valid. It sounds like there are no big differences from validity proofs. That's true. But the point is
this is used to prove the state transition `S` to `S''` is wrong by showing a valid state transition `S` to `S'`.

## ZKEVM Proofs

In the context of the EVM, which is regarded as a Turing machine, the generation of proofs requires demonstrating the
execution of Opcode operations in arbitrary order. These proofs serve as evidence of computation for the EVM.

> Additionally, validating the block data's integrity involves proving sub-operations such as Merkle Patricia Trie

Reviewer 2: Could you add the link of MPT in `glossary.md` to `Merkle Patricia Trie`?

Author: applied.

and ECDSA verification.
To support the proof generation process for target blocks, [zkevm-circuits] offer circuits capable of proving the
required sub-operations. These circuits provide the means to prove the integrity and validity of the block data.

> By combining these circuits, the Prover completes the overall validity proof for the target block.  

Reviewer 2: It would be better to add an empty line here to improve readability.

Author: applied

See [zkevm-specs] for details about the statements claimed by the zk-proof.

## The ZK Verifier Contract

The proofs generated by the Prover can be verified through a Verifier contract that includes the following interface.

> The challenge process is implemented using the `verify` function provided by the Verifier contract.

Reviewer 2: ```suggestion
The verification in challenge process is implemented using the `verify` function provided by the Verifier contract.
```

Author: applied

}
```

## Prover as an RPC Server

> [kroma-prover](https://github.com/kroma-network/kroma-prover) is implemented as a gRPC server that generates zkevm
> proofs for the requested height block. It can be utilized as a local component on the Challenger's machine

Reviewer: ```suggestion
[kroma-prover](https://github.com/kroma-network/kroma-prover) is implemented as a gRPC server that generates a zkevm
proof for the requested height block. It can be utilized as a local component on the challenger's machine
```

Author: applied

or as a remote [gRPC] server (prover as a service).

> - Step 1:  The user (e.g., Challenger) requests a zkevm-proof for a block of specific height to kroma-prover with
>   desired block height value.

Reviewer: ```suggestion
- Step 1:  An user (e.g., Challenger) requests a zkevm-proof for a block of specific height to kroma-prover with
  a desired block height value.
```

Author: applied

> - Step 2: The kroma-prover obtains the trace of the corresponding block from kroma-geth RPC.

Reviewer: ```suggestion
- Step 2: A kroma-prover obtains the trace of the corresponding block from kroma-geth RPC.
```

Author: applied

> - Step 3: And then the kroma-prover generates zkevm-proof for target block, and return it to the user.

Reviewer: ```suggestion
- Step 3: And then a kroma-prover generates zkevm-proof for the target block, and return it to the user.
```

Author: applied

> Operating the grpc-kroma-prover, which is launched only when necessary, can alleviate the situation where

Reviewer: ```suggestion
Operating a grpc-kroma-prover, which is launched only when necessary, can alleviate the situation where
```

Author: applied

}
```

## Prover as an RPC Server

[kroma-prover](https://github.com/kroma-network/kroma-prover) is implemented as a gRPC server that generates zkevm
proofs for the requested height block. It can be utilized as a local component on the Challenger's machine
or as a remote [gRPC] server (prover as a service).

- Step 1:  The user (e.g., Challenger) requests a zkevm-proof for a block of specific height to kroma-prover with
desired block height value.
- Step 2: The kroma-prover obtains the trace of the corresponding block from kroma-geth RPC.  

> - Step 3: And then the kroma-prover generates zkevm-proof for target block, and return it to the User.

Reviewer 2: ```suggestion
- Step 3: And then the kroma-prover generates zkevm-proof for target block, and return it to the user.
```

Author: applied

Operating the grpc-kroma-prover, which is launched only when necessary, can alleviate the situation where

> regular validators and challengers need to allocate excessive system resources due to the proof generation process
> that is occasionally executed (perhaps rarely executed)

Reviewer 2: ```suggestion
regular challengers need to allocate excessive system resources due to the proof generation process
that is occasionally executed (perhaps rarely executed).
```

Author: applied

### Protobuf

The detailed gRPC specification can be available at [prover-grpc-proto].

## Summary of Definitions

### Constants

| Name               | Value | Unit    |
|--------------------|-------|---------|

> | `FETCHING_TIMEOUT` | TBD   | seconds |

Reviewer 2: We do not have this constant in this doc.

Author: The value are not fixed.

Author: Are you mean that table would be removed?

Author: I agree with you. Once the content is defined, let's recreate the table.
